hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/DiskBalancerManager.java

```diff
@@ -0,0 +1,239 @@
+      reportList.add(HddsProtos.DatanodeDiskBalancerInfoProto.newBuilder()
+          .setCurrentVolumeDensitySum(String.valueOf(volumeDensitySum))
+          .setNode(datanodeDetails.toProto(clientVersion))
+          .build());
+    }
+
+    reportList.sort((t1, t2) -> t2.getCurrentVolumeDensitySum().
+        compareTo(t1.getCurrentVolumeDensitySum()));
+    return reportList.stream().limit(count).collect(Collectors.toList());
+  }
+
+  public List<HddsProtos.DatanodeDiskBalancerInfoProto> getDiskBalancerStatus(
+      Optional<List<String>> hosts, int clientVersion) throws IOException {
+    List<HddsProtos.DatanodeDiskBalancerInfoProto> statusList =
+        new ArrayList<>();
+    List<DatanodeDetails> filterDns = null;
+
+    if (hosts.isPresent() && !hosts.get().isEmpty()) {
+      filterDns = mapHostnamesToDatanodes(hosts.get());
+    }
+
+    for (DatanodeDetails datanodeDetails: nodeManager.getNodes(IN_SERVICE,
+        HddsProtos.NodeState.HEALTHY)) {
+      if ((filterDns != null && !filterDns.isEmpty() &&
+          filterDns.contains(datanodeDetails) &&
+          statusMap.containsKey(datanodeDetails)) ||
+          isRunning(datanodeDetails)) {
+        double volumeDensitySum =
+            getVolumeDataDensitySumForDatanodeDetails(datanodeDetails);
+        statusList.add(HddsProtos.DatanodeDiskBalancerInfoProto.newBuilder()
+            .setCurrentVolumeDensitySum(String.valueOf(volumeDensitySum))
+            .setDiskBalancerRunning(isRunning(datanodeDetails))
+            .setDiskBalancerConf(statusMap.get(datanodeDetails)
+                .getDiskBalancerConfiguration().toProtobufBuilder())
+            .setNode(datanodeDetails.toProto(clientVersion))
+            .build());
+      }
+    }
+
+    return statusList;
+  }
+
+  /**
+   * Get volume density for a specific DatanodeDetails node.
+   *
+   * @param datanodeDetails DatanodeDetails
+   * @return DiskBalancer report.
+   */
+  private double getVolumeDataDensitySumForDatanodeDetails(
+      DatanodeDetails datanodeDetails) {
+    Preconditions.checkArgument(datanodeDetails instanceof DatanodeInfo);
+
+    DatanodeInfo datanodeInfo = (DatanodeInfo) datanodeDetails;
+
+    double totalCapacity = 0d, totalUsed = 0d;
+    for (StorageContainerDatanodeProtocolProtos.StorageReportProto reportProto :
+        datanodeInfo.getStorageReports()) {
+      totalCapacity += reportProto.getCapacity();
+      totalUsed += reportProto.getScmUsed();
+    }
+
+    Preconditions.checkArgument(totalCapacity != 0);
+    double idealUsage = totalUsed / totalCapacity;
+
+    double volumeDensitySum = datanodeInfo.getStorageReports().stream()
+        .map(report ->
+            Math.abs((double)report.getScmUsed() / report.getCapacity()
+                - idealUsage))
+        .mapToDouble(Double::valueOf).sum();
+
+    return volumeDensitySum;
+  }
+
+  private List<DatanodeDetails> mapHostnamesToDatanodes(List<String> hosts)
```

Comment:
Please move this function to some utility class, so that disk balancer and node decommission can both use it.

Reply:
Updated.